### PR TITLE
Add telemetry for `SourceBuffer.appendBuffer`

### PR DESCRIPTION
**Pull request by at-ninja**
Modifies the SourceBufferMetrics class to not be completely stubbed out in gold builds, and use it to send UMA histograms for the time spent in `SourceBuffer::PrepareAppend` and
`SourceBufferAppendAlgorithm::Process`.

Added basic unit tests for the `SourceBufferMetrics` class, to assert that the new code reports telemetry against the correct histograms.

b/329439818

Change-Id: I7921434b02c689b645ab2a60464eee166d829135

### CODE REVIEW

File: cobalt/dom/source_buffer_metrics.h [hunk 1]
```diff
@@ -15,58 +15,63 @@
...
  public:
-  explicit SourceBufferMetrics(bool is_primary_video)
-      : is_primary_video_(is_primary_video) {}
+  SourceBufferMetrics(
+      bool is_primary_video,
+      const base::TickClock* clock = base::DefaultTickClock::GetInstance())
+      : is_primary_video_(is_primary_video), clock_(clock) {}
   ~SourceBufferMetrics() = default;
 
   void StartTracking();
```
Comment (sideb0ard) [inline cobalt/dom/source_buffer_metrics.h:39]: lets add the SourceBufferMetricsAction enum to StartTracking too, to be consistent with EndTracking. Lets add a DCHECK or similar to ensure the EndTracking matches the correct StartTracking

**Review by sideb0ard — CHANGES_REQUESTED** [review]
Looks great! Added some minor comments.

File: tools/metrics/histograms/metadata/cobalt/histograms.xml [hunk 2]
```diff
@@ -31,6 +31,13 @@ Always run the pretty print utility on this file after editing:
 
 <histograms>
 
+<variants name="SourceBufferMetricsAction">
```
[the remark this answers is not in the document]
Comment (joeltine) [reply tools/metrics/histograms/metadata/cobalt/histograms.xml:34]: Correct. Sorry about that. Just haven't implemented this yet.

File: cobalt/dom/source_buffer_metrics.cc [hunk 3]
```diff
@@ -63,44 +63,89 @@ double GetWallToThreadTimeRatio(int64_t wall_time, int64_t thread_time) {
...
+}
+
+void SourceBufferMetrics::RecordTelemetry(
+    SourceBufferMetricsAction action, const base::TimeDelta& action_duration) {
+  switch (action) {
+    case SourceBufferMetricsAction::PREPARE_APPEND:
+      UMA_HISTOGRAM_CUSTOM_MICROSECONDS_TIMES(
+          "Cobalt.Media.SourceBuffer.PrepareAppend.Timing", action_duration,
+          base::TimeDelta::FromMicroseconds(1), base::TimeDelta::FromSeconds(1),
+          50);
```
Comment (joeltine) [inline cobalt/dom/source_buffer_metrics.cc:121]: Were you able to test these histograms locally? How was your bucket distribution? I wonder about your min/max ranges here and your bucket size (50). 1 microsecond to 1 second is a delta of 999,999. This is a very wide range... is 50 buckets too coarse? Can you tighten your interval here reasonably and also perhaps up the buckets to 100?

**Review by sideb0ard — APPROVED** [review]
thanks for the changes, lgtm, pending Joel's approval of histogram buckets.

File: tools/metrics/histograms/metadata/cobalt/histograms.xml [hunk 4]
```diff
@@ -151,6 +151,40 @@ Always run the pretty print utility on this file after editing:
   </summary>
 </histogram>
 
+<histogram name="Cobalt.Media.SourceBuffer.AppendBuffer.Timing"
+    units="ms" expires_after="never">
```
Comment (joeltine) [inline tools/metrics/histograms/metadata/cobalt/histograms.xml:155]: aren't your units microseconds?

File: cobalt/dom/source_buffer_metrics.cc (hunk 3 above)
Comment (joeltine) [reply cobalt/dom/source_buffer_metrics.cc]: Looks good.

File: tools/metrics/histograms/metadata/cobalt/histograms.xml (hunk 4 above)
Comment (at-ninja) [reply tools/metrics/histograms/metadata/cobalt/histograms.xml]: Fixed.

**Review by joeltine — APPROVED** [review]

**Pull request merged**